tests/signals/tests.py

```diff
@@ -351,3 +351,24 @@ def test_register_model_class_senders_immediately(self):
         apps2 = Apps()
         signals.post_init.connect(self.receiver, sender=Book, apps=apps2)
         self.assertEqual(list(apps2._pending_operations), [])
+
+
+class ModelStateTests(TestCase):
+    databases = ['default', 'other']
+
+    def setUp(self):
+        a1 = Author.objects.using('other').create(id=1)
+        b1 = Book.objects.using('other').create(id=1)
+        b1.authors.set([a1])
+
+    def test_model_state(self):
+        def post_signal(instance, **kwargs):
+            import copy
+            self.instance = copy.deepcopy(instance)
+
+        signals.post_init.connect(post_signal, sender=Book)
+        Book.objects.using('other').get(pk=1)
+        self.assertFalse(self.instance._state.adding)
+        self.assertEqual(self.instance._state.db, 'other')
+        self.assertEqual(self.instance._state.db, self.instance.authors.get()._state.db)
+        signals.post_init.disconnect(post_signal, sender=Book)
```

Comment:
It feels cleaner if you would do this during cleanup of the test, basically doing:

```python
def test_model_state(self):
    def disconnect():
        signals.post_init.disconnect(post_signal, sender=Book)

    signals.post_init.connect(post_signal, sender=Book)
    self.addCleanup(disconnect)

    ...
```

Because this will cause the signal to disconnect even if an assertion fails.

Reply:
@Tirzono `pre_init` don't take the `instance` parameter. I don't think so if we need `pre_init` test.

Reply:
Ah ofcourse, my bad, you're right.

Looks good to me.